### PR TITLE
Allow for throttling of requests

### PR DESCRIPTION
**Pull request by gabrielwong**

### CODE REVIEW

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala [hunk 1]
```diff
@@ -345,11 +358,30 @@ trait PartitionedNetworkClient[PartitionedId] extends BaseNetworkClient {
     if (nodes.size <= 1 || !routingConfigs.selectiveRetry || retryStrategy == None) {
       val queue = new ResponseQueue[ResponseMsg]
       val resIter = new NorbertDynamicResponseIterator[ResponseMsg](nodes.size, queue)
+      lazy val sem = new Semaphore(maxConcurrentRequests)
+
+      val addResult = (res: Either[Throwable, ResponseMsg]) => {
+        queue += res
+        if (maxConcurrentRequests > 0)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: Please use curly braces

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala [hunk 2]
```diff
@@ -345,11 +358,30 @@ trait PartitionedNetworkClient[PartitionedId] extends BaseNetworkClient {
...
-          doSendRequest(PartitionedRequest(requestBuilder(node, idsForNode), node, idsForNode, requestBuilder, is, os, if (maxRetry == 0) Some((a: Either[Throwable, ResponseMsg]) => {queue += a :Unit}) else Some(retryCallback[RequestMsg, ResponseMsg](queue.+=, maxRetry, capability, persistentCapability)_), 0, Some(resIter)))
+          if (maxConcurrentRequests > 0)
+            sem.acquire()
+          doSendRequest(PartitionedRequest(requestBuilder(node, idsForNode), node, idsForNode, requestBuilder, is, os,
+              callback, 0, Some(resIter)))
         } catch {
-          case ex: Exception => queue += Left(ex)
+          case ex: Exception => {
+            queue += Left(ex)
+            if (maxConcurrentRequests > 0)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: More curly braces

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala [hunk 3]
```diff
@@ -345,11 +358,30 @@ trait PartitionedNetworkClient[PartitionedId] extends BaseNetworkClient {
...
+      }
+      val callback = if (maxRetry == 0)
+        Some(addResult)
+      else
+        Some(retryCallback[RequestMsg, ResponseMsg](addResult, maxRetry, capability, persistentCapability)_)
+
       nodes.foreach { case (node, idsForNode) =>
         try {
-          doSendRequest(PartitionedRequest(requestBuilder(node, idsForNode), node, idsForNode, requestBuilder, is, os, if (maxRetry == 0) Some((a: Either[Throwable, ResponseMsg]) => {queue += a :Unit}) else Some(retryCallback[RequestMsg, ResponseMsg](queue.+=, maxRetry, capability, persistentCapability)_), 0, Some(resIter)))
+          if (maxConcurrentRequests > 0)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: More curly braces

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala [hunk 4]
```diff
@@ -345,11 +358,30 @@ trait PartitionedNetworkClient[PartitionedId] extends BaseNetworkClient {
...
       val queue = new ResponseQueue[ResponseMsg]
       val resIter = new NorbertDynamicResponseIterator[ResponseMsg](nodes.size, queue)
+      lazy val sem = new Semaphore(maxConcurrentRequests)
+
+      val addResult = (res: Either[Throwable, ResponseMsg]) => {
+        queue += res
+        if (maxConcurrentRequests > 0)
+          sem.release()
+      }
+      val callback = if (maxRetry == 0)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: Curlies

File: gradle.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
-version=0.6.92
+version=0.6.93
```
Comment (sungjuc) [inline gradle.properties]: You may want to rebase the respository since current version is 93.

**Pull request closed without merging**